File: persistence-test/src/test/java/org/mariella/test/model/Entity.java
package org.mariella.test.model;

import javax.persistence.Column;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import javax.persistence.Transient;
import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.util.UUID;

@MappedSuperclass
public class Entity {
    @Transient
    protected final PropertyChangeSupport propertyChangeSupport = new PropertyChangeSupport(this);
    private UUID id;

    public Entity() {
        this(UUID.randomUUID());
    }

    public Entity(UUID id) {
        this.id = id;
    }

    public void addPropertyChangeListener(PropertyChangeListener listener) {
        propertyChangeSupport.addPropertyChangeListener(listener);
    }

    public void removePropertyChangeListener(PropertyChangeListener listener) {
        propertyChangeSupport.removePropertyChangeListener(listener);
    }

    @Id
    @Column(name = "ID")
//	@GeneratedValue(strategy=GenerationType.AUTO)
    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        propertyChangeSupport.firePropertyChange("id", this.id, this.id = id);
    }

}

File: persistence-kotlin/src/test/kotlin/org/mariella/test/entities/Entity.kt
package org.mariella.test.entities

import org.mariella.persistence.kotlin.TrackingSupport
import java.time.Instant
import java.util.*
import javax.persistence.Column
import javax.persistence.Id
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class Entity : TrackingSupport() {
    companion object {
        private const val MAX_DB_TIMESTAMP_MILLIS = 253402214400000
        val MAX_DB_TIMESTAMP: Instant = Instant.ofEpochMilli(MAX_DB_TIMESTAMP_MILLIS)
    }

    @get:Id
    @get:Column(name = "id")
    var id: UUID by changeSupport(UUID.randomUUID())
}